components/datetime/src/pattern/mod.rs

```diff
@@ -281,3 +288,113 @@ impl Serialize for Pattern {
         }
     }
 }
+
+/// Used to represent either H11/H12, or H23/H24. Skeletons only store these
+/// hour cycles as H12 or H24.
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub enum FlexibleHourCycle {
+    /// Can either be fields::Hour::H11 or fields::Hour::H12
+    H11H12,
+    /// Can either be fields::Hour::H23 or fields::Hour::H24
+    H23H24,
+}
+
+/// Default is required for serialization, arbitrarily pick one.
+impl Default for FlexibleHourCycle {
+    fn default() -> Self {
+        FlexibleHourCycle::H11H12
+    }
+}
+
+impl FlexibleHourCycle {
+    pub fn matches_field(&self, hour: &fields::Hour) -> bool {
+        match self {
+            FlexibleHourCycle::H11H12 => match hour {
+                fields::Hour::H11 | fields::Hour::H12 => true,
+                fields::Hour::H23 | fields::Hour::H24 => false,
+            },
+            FlexibleHourCycle::H23H24 => match hour {
+                fields::Hour::H11 | fields::Hour::H12 => false,
+                fields::Hour::H23 | fields::Hour::H24 => true,
+            },
+        }
+    }
+
+    pub fn matches_preference(&self, hour: &preferences::HourCycle) -> bool {
+        match self {
+            FlexibleHourCycle::H11H12 => match hour {
+                preferences::HourCycle::H11 | preferences::HourCycle::H12 => true,
+                preferences::HourCycle::H23 | preferences::HourCycle::H24 => false,
+            },
+            FlexibleHourCycle::H23H24 => match hour {
+                preferences::HourCycle::H11 | preferences::HourCycle::H12 => false,
+                preferences::HourCycle::H23 | preferences::HourCycle::H24 => true,
+            },
+        }
+    }
+
+    /// Skeletons only use h12 and h24.
+    fn to_skeleton_field_symbol(&self) -> fields::Hour {
+        match self {
+            FlexibleHourCycle::H11H12 => fields::Hour::H12,
+            FlexibleHourCycle::H23H24 => fields::Hour::H24,
+        }
+    }
+}
+
+pub fn determine_flexible_hour_cycle(pattern: &Pattern) -> Option<FlexibleHourCycle> {
+    for item in pattern.items() {
+        if let PatternItem::Field(fields::Field { symbol, length: _ }) = item {
+            if let fields::FieldSymbol::Hour(pattern_hour) = symbol {
+                return Some(match pattern_hour {
+                    fields::Hour::H11 | fields::Hour::H12 => FlexibleHourCycle::H11H12,
+                    fields::Hour::H23 | fields::Hour::H24 => FlexibleHourCycle::H23H24,
+                });
+            }
+        }
+    }
+
+    None
+}
+
+/// Invoke the pattern matching machinery to transform the hour cycle of a pattern. This provides
+/// a safe mapping from a h11/h12 to h23/h24
+pub fn apply_flexible_hour_cycle(
```

Comment:
Issue: I think this method is only used in the transformer.  I'd like to explore whether we can get code used only for the transformer out of the core library, or at least behind a feature flag.  I'd be interested in your thoughts on this.

Reply:
Adding a `provider_transform_utils` feature flag seems like a reasonable-enough way to hide this from end-users, but still allow us to use internals. The inter-dependencies seem like they could get complicated quick, so a new feature flag seems reasonable. It's nice being able to leverage all of pattern and skeleton utilities to make this stuff work well.